### PR TITLE
max_cluster_size parameter

### PR DESCRIPTION
**Pull request by Rocketknight1**
Hi, this is the PR for issue #408 ! I've done some quick testing and it seems to be working correctly - the code changes are quite minor. I'm not 100% sure I'm correctly computing the 'size' of clusters - you can see I'm counting the size of a cluster as the 'child_size' of the edge that it's a child of, but I don't know if that incorrectly counts some points that have already fallen out of that cluster. Would it be more correct to count the size of non-leaf clusters as the sum of the child_size of the two clusters branching off from it?

One other notable thing is that I haven't yet edited the flat.py code - I wasn't sure if a max_cluster_size parameter made sense there, and I don't understand that bit of the codebase that well. All the functions where max_cluster_size is used have a default value of 0 (i.e. no max size) set, so all other code should be unaffected by this change, even if they're still expecting the old function signature.

### CODE REVIEW

**Comment by lmcinnes** [comment]
Thanks for this. It looks good at a glance. I'll try to get some time to go through it in more detail soon.

**Comment by Rocketknight1** [comment]
Hey, just bumping this so it doesn't get lost!

**Comment by Rocketknight1** [comment]
Hey, update - although this looks like it works at first glance, it's actually quite wrong - I think the max_cluster_size parameter gets stashed in the `_metric_kwargs` dict, which isn't what we want. I got away with it because those kwargs get expanded in the call to the `hdbscan()` method, but it causes problems if we do other things like call `generate_prediction_data()`.

I'll try to fix and update the PR!

**Comment by Rocketknight1** [comment]
Updated with a fix that looks like it's working, but I'll keep poking at it

**Comment by lmcinnes** [comment]
I'm sorry I haven't had time to look at this properly. I'll try to look at it soon, but I have some other matters to deal with as well that may take up most of my time. Again, sorry about that.

**Comment by Rocketknight1** [comment]
Further update: The max_cluster_size parameter is not being handled correctly when we want to predict new points using approximate_predict or membership_vector, and so predictions are made on the set of cluster labels that would have existed if there was no max cluster size, which isn't that helpful.

There may be some other 'traps' like this in the code that I'm not aware of, but I'll dig in and see if I can figure out how to resolve this one, at least.

**Comment by Rocketknight1** [comment]
Update: The root cause seems to be that when prediction data is generated, cluster selection is [performed on the condensed tree entirely independently](https://github.com/scikit-learn-contrib/hdbscan/blob/master/hdbscan/prediction.py#L106-L111) from the selection that's performed to generate labels - presumably this works because with the same tree and choice of cluster_selection_method, you generally get the same results.

However, I think this is problematic - it uses [CondensedTree._select_clusters()](https://github.com/scikit-learn-contrib/hdbscan/blob/master/hdbscan/plots.py#L234), whereas elsewhere the [Cython function get_clusters()](https://github.com/scikit-learn-contrib/hdbscan/blob/master/hdbscan/_hdbscan_tree.pyx#L659) is used. This means that parameters like cluster_selection_epsilon (and now max_cluster_size) seem to be ignored when using approximate_predict() or membership_vector().

I'll do some testing to confirm that cluster_selection_epsilon is also not propagated correctly when using approximate_predict().

**Comment by Rocketknight1** [comment]
I've confirmed that `approximate_predict()` predicts based as if cluster_selection_epsilon had not been used - attached is a quick interactive session showing that it predicts cluster labels for new points that are outside the range that exist in the training data. 

```
>>> import numpy as np
>>> from hdbscan import HDBSCAN
>>> train_data = np.random.normal(size=(10000, 10))
>>> test_data = np.random.normal(size=(10000, 10))
>>> clusterer = HDBSCAN(min_samples=2, min_cluster_size=10, cluster_selection_method='leaf', prediction_data=True)
>>> labels = clusterer.fit_predict(train_data)
>>> np.max(labels)
18
>>> clusterer = HDBSCAN(min_samples=2, min_cluster_size=10, cluster_selection_method='leaf', prediction_data=True, cluster_selection_epsilon=1.2)
>>> labels = clusterer.fit_predict(train_data)
>>> np.max(labels)
12
>>> from hdbscan.prediction import approximate_predict
>>> out = approximate_predict(clusterer, test_data)
>>> out
(array([-1, -1, -1, ..., -1, -1, -1]), array([0., 0., 0., ..., 0., 0., 0.]))
>>> np.max(out[0])
18
```

**Comment by Rocketknight1** [comment]
I think this indicates the need for a refactor where cluster data isn't generated twice, or at least that identical function calls and parameters are used if it is. Let me know if you want me to try to incorporate that into this PR!

**Comment by lmcinnes** [comment]
In the end a great deal of bits and pieces agglomerated, including the prediction and soft clustering, which, as you've noted, ultimately causes issues because things have been quasi-duplicated in places. At this point I find the prediction somewhat secondary, and the soft clustering is largely deprecated at this point. At this point I think the easiest thing to do is add some notes in the docstrings that ``max_cluster_size`` and ``cluster_epsilon`` do not play well with prediction. That would allow us to get this PR merged, and then worry about the required refactoring later. So beyond those docstring notes the current changes look good to me. If you want to add some notes we can get this merged.

**Comment by Rocketknight1** [comment]
Hi, I'm sorry for the very long delay here! I was experimenting locally to see if I could find a good solution, but I didn't have a lot of time because of other work stuff, so I didn't get too far. In the end I added the docstrings as you suggested.

**Comment by Rocketknight1** [comment]
If you're happy with the docstrings we can merge as is.

**Comment by lmcinnes** [comment]
Seems like a strange error in the boruvka ball tree in the tests, but it may be just a bit of noise (it works most of the time). Thanks for this, I'll just merge it and deal with any remaining issues later. I know this took a long time, so thanks for the persistence.

**Comment by Rocketknight1** [comment]
Don't worry about it, a lot of the delay was just me not doing anything with it. Thank you!
